src/compiler/checker.ts

```diff
@@ -9982,6 +9982,26 @@ namespace ts {
             if (checkType === wildcardType || extendsType === wildcardType) {
                 return wildcardType;
             }
+            // Simplifications for types of the form `T extends U ? U : never` and `T extends U ? never : T`.
+            const originalCheckType = getActualTypeVariable(root.checkType);
+            const trueType = instantiateType(root.trueType, mapper);
+            const falseType = instantiateType(root.falseType, mapper);
+            if (falseType.flags & TypeFlags.Never && getActualTypeVariable(root.trueType) === originalCheckType) {
+                if (isTypeAssignableTo(getRestrictiveInstantiation(checkType), getRestrictiveInstantiation(extendsType))) { // Always true
+                    return getIntersectionType([trueType, extendsType]);
+                }
+                else if (getUnionType([getIntersectionType([checkType, extendsType]), neverType]).flags & TypeFlags.Never) { // Always false

```

Comment:
Shouldn't this check the permissive form of the check and extends types? I'm not really following the logic here.

Reply:
So, the permissive form check returns `true` for `X extends never`, while this version of the check does not. We want to ignore the never extends never case here, since that will become a `never` result in both branches anyway (meaning we'd like x extends never of x extends something instantiable to never to simplify even when the permissive instantiation says there's technically one instantiatiom which'd give another result).

With negated types in place, I'd probably write the check as `isTypeAssignableTo(checkType, getNegatedType(extendsType))` instead, but without that in place, this accomplishes the same thing when only normal types exist.